src/claimtriedb.h

```diff
@@ -0,0 +1,140 @@
+#ifndef BITCOIN_CLAIMTRIE_DB_H
+#define BITCOIN_CLAIMTRIE_DB_H
+
+#include "dbwrapper.h"
+
+#include <utility>
+
+#include <map>
+
+/**
+ * additional macros to complete hash algorithm
+ */
+#define  len(s)          (sizeof(s)/sizeof(*s)-1)
+#define   H1(s, i, x)    (x*65599u+(uint8_t)s[(i)<len(s)?len(s)-1-(i):len(s)])
+#define   H4(s, i, x)  H1(s, i,  H1(s, i+ 1,  H1(s, i+  2,  H1(s, i+  3, x))))
+#define  H16(s, i, x)  H4(s, i,  H4(s, i+ 4,  H4(s, i+  8,  H4(s, i+ 12, x))))
+#define  H64(s, i, x) H16(s, i, H16(s, i+16, H16(s, i+ 32, H16(s, i+ 48, x))))
+#define H256(s, i, x) H64(s, i, H64(s, i+64, H64(s, i+128, H64(s, i+192, x))))
+
```

Comment:
Where did these definitions come from?  It should be documented because 1) Not all hashes are equally appropriate, and 2) I'd hope we can replace most/all of these with boost:: or std:: hash implementations.  I don't think supporting these in our code forever is ideal.

Reply:
Get it from here -> http://lolengine.net/blog/2011/12/20/cpp-constant-string-hash it *should* be unique.

Reply:
This link didn't work for me.  It's also not really documenting why we want to use these hash values in our code base indefinitely, as opposed to re-organizing or sticking with existing boost/std ones (or some composition of).

Reply:
@bvbfan , can you explain why these hash methods weren't necessary before but are now?

Reply:
Sure. Whole idea behind these hashes is to use as safe as possible db. Till now they a predefined chars 'n', 'r' etc. which was used in claimtrie.cpp now we separated it and calls can be type safe e.g. you can only declare type and don't care what is predefined char, hash or anything else. The first idea was to use boost::type_id but for sake of not generate same hashes in all its versions (e.g. in 1.59 and 1.64 are not same) i aim to provide a proven algorithm to generate same hashes in all compilers and OS'es. That's why unit test matters to prove any type that we need has a unique value.

Reply:
Not only this, generic functions prevent code duplication and it does not need to update db implementation every time when new type is introduced. I've notice many code duplication in claimtrie i aim to remove it at all, this kind of implementation do it.